### PR TITLE
WebPart for sending streams

### PR DESCRIPTION
**Pull request by njlr**
- Add a WebPart for sending streams (`okStream`)
 - Add an example of sending a stream

### CODE REVIEW

**Comment by ademar** [comment]
Hi, thanks for the PR. 
 
I think for the feature to be completed it should correctly handle the case where the length of the stream is unknown.
In that case it we should transmit the stream down the pipe using chunked transfer encoding.

If you handle that case and add some unit tests we can merge your work.

Thanks,

**Comment by njlr** [comment]
> Hi, thanks for the PR.
> 
> I think for the feature to be completed it should correctly handle the case where the length of the stream is unknown. In that case it we should transmit the stream down the pipe using chunked transfer encoding.
> 
> If you handle that case and add some unit tests we can merge your work.
> 
> Thanks,

Great!

I have added a chunked variant and also some tests. Please let me know what you think 👍

File: src/Suave/Stream.fs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+              status = HTTP_200.status
+              content = SocketTask write
+          }
+    }
+    |> succeed
+
+/// Send a stream back in the response with 200 status using chunked transfer-encoding.
+/// A new stream will be created for every request and it will be disposed after the request completes.
+/// You are responsible for setting the MIME type.
+let okStreamChunked (makeStream : Async<Stream>) : WebPart =
```
Comment (ademar) [inline src/Suave/Stream.fs:46]: Same here, use `Stream` rather than `Async<Stream>`

File: src/Suave/Stream.fs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  fun ctx ->
+    let write (conn, _) =
+      socket {
+        use! stream = SocketOp.ofAsync makeStream
+
+        if ctx.request.``method`` <> HttpMethod.HEAD then
+          do! transferStreamChunked conn stream
```
Comment (ademar) [inline src/Suave/Stream.fs:53]: Shouldn't we send a newline here before sending the stream ?

**Review by ademar — CHANGES_REQUESTED** [review]
Hi, I've made some suggestions that I think make sense. Let me know what you think.

File: src/Suave/Stream.fs [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+open System.IO
+open Suave
+open Suave.Sockets
+open Suave.Sockets.Control
+
+/// Send a stream back in the response with 200 status.
+/// A new stream will be created for every request and it will be disposed after the request completes.
+/// You are responsible for setting the MIME type.
+/// The stream must support the `Length` property.
+let okStream (makeStream : Async<Stream>) : WebPart =
```
[the remark this answers is not in the document]
Comment (njlr) [reply src/Suave/Stream.fs:12]: The reason I did it this way is to address a few things:

 1. The exact stream we need may only be known after inspecting the request and calling another service. For example, imagine we look into the request to get the user ID and then hit another service to fetch their profile picture. This process will likely be `Async`.
 
 2. Usually, we want to dispose of a stream when we are finished with it. I couldn't come up with a design that triggers disposal correctly without giving each invocation of the `WebPart` ownership of the stream
 
 3. Most streams do not allow multiple concurrent users, and giving the `WebPart` a mechanism for requesting a new stream instance side-steps many problems related to this

[the remark this answers is not in the document]
Comment (ademar) [reply src/Suave/Stream.fs:12]: OK, let's keep it like this.

**Comment by njlr** [comment]
> Hi, I've made some suggestions that I think make sense. Let me know what you think.

I have pushed the suggested changes.

**Review by ademar — APPROVED** [review]

**Comment by ademar** [comment]
Thank you very much!

**Pull request merged**